core/services/major_tom/src/core.py

```diff
@@ -0,0 +1,148 @@
+        return float(ping)
+    except Exception:
+        return -1.0
+
+
+@dataclass
+class DefaultPayload:
+    log_session_uuid: str
+    order: int
+    timestamp: str
+    hardware_id: str
+    blueos_id: str
+    data: dict[str, Any]
+
+    def json(self) -> dict[str, Any]:
+        return asdict(self)
+
+
+class TelemetryEngine:
+    # pylint: disable=too-many-arguments
+    def __init__(
+        self,
+        label: str,
+        endpoint: str,
+        s3_endpoint: str,
+        create_record: Callable[[Any], Any],
+        interval: float,
+        max_file_size: int,
+        max_file_retention: int,
+        buffer_folder: str,
+        log_buffer: loguru._logger.Logger,  # type: ignore
+    ):
+        self.buffer_file = f"{buffer_folder}/{label}_usage.log"
+        self.buffer_folder = buffer_folder
+
+        self.telemetry_endpoint = endpoint
+        self.telemetry_s3_endpoint = s3_endpoint
+        self.create_record = create_record
+        self.interval = interval
+
+        self.log_buffer = log_buffer
+        self.log_buffer.add(
+            self.buffer_file,
+            rotation=max_file_size,
+            retention=max_file_retention,
+            format=formatter,
+            compression="gz",
+        )
+
+    def __call__(self) -> None:
+        order = 0
+        while True:
+            order += 1
+            record = self.create_record(order)
+            if self.save(record) == "online":
+                self.process_buffered_records()
+            time.sleep(self.interval)
+
+    def upload_file(self, file: str) -> bool:
+        """
+        This method requests to telemetry API a presigned url and upload the local archived files.
+        """
+        logger.info(f"uploading file... {file}")
+        try:
+            response = requests.get(self.telemetry_s3_endpoint, timeout=5).json()
+            with open(file, "rb") as fh:
+                files = {"file": (file, fh)}
+                r = requests.post(response["url"], data=response["fields"], files=files, timeout=300)
+                if r.status_code == 204:
+                    logger.info("[Success!]")
+                    return True
+        except Exception as error:
+            logger.info("Ground Control to Major Tom. Your circuit's dead, there's something wrong.")
+            logger.error(f"error upload log file: {error}")
+
+        return False
+
+    def process_buffered_records(self) -> None:
+        """
+        Check in the buffered folder if there are archived logs to upload. If the agent connects before an archive
+        is created it will also archive the current buffer file and upload it.
+        """
+        for file in os.listdir(self.buffer_folder):
+            file_path = os.path.join(self.buffer_folder, file)
+
+            # Upload regular archive
+            if file_path.endswith(".log.gz"):
+                if self.upload_file(file_path):
+                    os.remove(file_path)
```

Comment:
I believe that there is an issue on the logic, if the file is already .gz, there is no reason to run the following lines
```suggestion
            if file_path.endswith(".log.gz") and self.upload_file(file_path):
                os.remove(file_path)
                continue
```